File: 0623-add-one-row-to-tree/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    void levelorder(TreeNode*& n, int v, int d)
    {
        if(n == nullptr)
        {
            n = new TreeNode(v);
            return;
        }
        else
        {
            int level = 1;
            queue<TreeNode*> L;
            L.push(n);
            
            
            while(L.size() != 0)
            {
                queue<TreeNode*> temp;
                if(level != d - 1)
                {
                    while(L.size() != 0)
                    {
                        TreeNode* node = L.front();
                        L.pop();
                        if(node->left != nullptr)
                            temp.push(node->left);
                        if(node->right != nullptr)
                            temp.push(node->right);
                    }
                    level++;
                    L = temp;
                }
                else
                {
                    while(L.size() != 0)
                    {
                        TreeNode*node = L.front();
                        L.pop();
                        
                        TreeNode* n_left = new TreeNode(v);
                        n_left->left = node->left;
                        
                        TreeNode* n_right = new TreeNode(v);
                        n_right->right = node->right;
                        
                        node->left = n_left;
                        node->right = n_right;  
                    }
                    
                    break;
                }
            }
            
            
        }
    }
public:
    TreeNode* addOneRow(TreeNode* root, int v, int d) {
        if(d == 1)
        {
            TreeNode* r = new TreeNode(v);
            r->left = root;
            r->right = nullptr;
            return r;
        }
        levelorder(root, v, d);
        return root;
    }
};

File: 0623-add-one-row-to-tree/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def addOneRow(self, root: Optional[TreeNode], val: int, depth: int) -> Optional[TreeNode]:
        if depth == 1:
            newRoot = TreeNode(val, root, None)
            return newRoot
        
        
        Q = collections.deque()
        Q.append(root)
                
        for curDepth in range(1, depth-1):
            newQ = collections.deque()            
            while Q:
                node = Q.popleft()
                if node.left:
                    newQ.append(node.left)
                if node.right:
                    newQ.append(node.right)
            
            Q = newQ
        
        for node in Q:
            left  = node.left
            right = node.right 
            node.left = TreeNode(val, left, None)
            node.right = TreeNode(val,None , right)

        return root
            
                
            
            
                
        
                
            
            
            
            
